Exceed Covariance Norm

samuelqian01 opened the pull request:
Modified Attitude_estimator ptest case and supporting files for a scenario that has been constructed (nothing written inside) for exceeding covariance norm. Added logic in attitude_estimator suite of files to trigger a reset when state field attitude_estimator_reset_persistance_reached becomes true (which triggers exceed_persistence), which could first be tested by manually triggering it in the ptest case but then adding the logic within attitude_estimator to trigger upon reaching some covariance.

afoarce commented on .github/workflows/hootl.yml:
```diff
@@ -56,6 +56,7 @@ jobs:
         !(python -m ptest runsim -c ptest/configs/hootl_speedup.json -t FailingEmptyCase -ni)
         !(python -m ptest runsim -c ptest/configs/hootl_speedup.json -t FailingEmptySimCase -ni)
         python -m ptest runsim -c ptest/configs/hootl.json -t AttitudeNoSunVectorsInitializationCase -ni
+        python -m ptest runsim -c ptest/configs/hootl.json -t AttitudePersistanceExceededCase - ni
```
remove space `-ni`

afoarce commented on src/fsw/FCCode/Estimators/AttitudeEstimator.cpp:
```diff
@@ -217,7 +221,7 @@ void AttitudeEstimator::_execute()
                     _state, time_s, {0.0f, 0.0f, 0.0f, 1.0f});
         }
-        else if (adcs_ssa_valid)
+        else if ((adcs_ssa_valid) || (exceed_persistance))
```
Is this the logic we want? a comment explaining the logic/reasoning would be helpful here

samuelqian01 replied:
After a closer look at the code I realized I just wrote everything that had already been written. I'm probably going to end up deleting most of the changes to attitude_estimator.cpp because I can leverage adcs_ssa_valid/change it to accommodate for accounting the Frobenius norm.

afoarce commented on install.sh:
```diff
@@ -0,0 +1,855 @@
+#!/bin/bash
```
don't think we need this file

knk38 commented on src/fsw/FCCode/Estimators/AttitudeEstimator.cpp:
```diff
@@ -43,6 +44,7 @@ AttitudeEstimator::AttitudeEstimator(StateFieldRegistry &registry)
       attitude_estimator_reset_cmd_f("attitude_estimator.reset_cmd", Serializer<bool>()),
       attitude_estimator_mag_flag_f("attitude_estimator.mag_flag", Serializer<bool>(), 1),
       attitude_estimator_ignore_sun_vectors_f("attitude_estimator.ignore_sun_vectors", Serializer<bool>(), 1),
+      attitude_estimator_reset_persistance_reached("attitude_estimator.reset_persistance_reached", Serializer<bool>(), 0),
```
why an eeprom save field value of zero here?

shihaocao commented on src/fsw/FCCode/Estimators/AttitudeEstimator.cpp:
```diff
@@ -197,6 +201,33 @@ void AttitudeEstimator::_execute()
     }();
     auto const b_body = attitude_estimator_b_body_f.get();
 
+    /* The Current Frobenius Norm */
+
+    float fro_norm= lin::fro(lin::nans<lin::Vector3d>());
+
+    /* Make sure to set the attitude_estimator_reset_persistance_reached to false check this, this is the logic
+    that triggers the state field to allow for a reset of attitude_estimator. Once you exceed baseline frobenius norm
+    * 1000 (for the safety factor), trigger the persistance state field, accounting for SSA valid and invalid. */
+    
+    if (fro_norm > 3.86e-7 * 1000 && !adcs_ssa_valid) {
+        attitude_estimator_reset_persistance_reached.set(true);
+    }
+
+    else if (fro_norm > 8.35e-10 * 1000 && adcs_ssa_valid) {
+        attitude_estimator_reset_persistance_reached.set(true);
+    }
+
+    auto const exceed_persistance = attitude_estimator_reset_persistance_reached.get();
+
+    /* One if statement for no ssa covariance and one if statement for ssa valid covariance */
+
+    if (!adcs_ssa_valid && exceed_persistance) {
+        gnc::attitude_estimator_reset(_state, time_s, {0.0f, 0.0f, 0.0f, 1.0f});
+    }
+    else if (adcs_ssa_valid && exceed_persistance) {
+        gnc::attitude_estimator_reset(_state, time_s, orbit_pos, b_body, adcs_ssa);
+    }
+
```
@govindchari

samuelqian01 commented:
<img width="1031" alt="Screen Shot 2021-10-16 at 6 08 34 PM" src="https://user-images.githubusercontent.com/65980474/137603305-85fc0311-2e8d-4c0a-90a0-94e1c161e2a3.png">
<img width="818" alt="Screen Shot 2021-10-16 at 6 08 22 PM" src="https://user-images.githubusercontent.com/65980474/137603307-ecda8567-3d0c-4d5d-bd73-9ca983fbdfb5.png">

We added persistence logic, created a writeable state field for the safety factor, and modified our ptest case to trigger the gnc::attitude_estimator_reset. The pictures above show the print statements we used to verify that a reset had occurred, and we tested by setting the safety factor extremely low to guarantee the triggering of exceed_persistance.

shihaocao commented:
We should close this PR as we unfortunately didn't have time to add this to the final FSW package.
